In [189]:
from google.colab import drive
drive.mount('/content/drive', force_remount = True)

Mounted at /content/drive


In [190]:
import pandas as pd
import numpy as np
import re

In [191]:
unfl_data =  pd.read_json("/content/drive/MyDrive/_Work@MachineLearning/Natural Language Processing/penn-data.json")

In [192]:
unfl_data

0  \
0     Pierre Vinken, 61 years old, will join the boa...   
1     Mr. Vinken is chairman of Elsevier N.V., the D...   
2     Rudolph Agnew, 55 years old and former chairma...   
3     A form of asbestos once used to make Kent ciga...   
4     The asbestos fiber, crocidolite, is unusually ...   
...                                                 ...   
3909  They also said that more than a dozen presiden...   
3910  Sen. Kennedy said in a separate statement that...   
3911  Trinity Industries Inc. said it reached a prel...   
3912                           Terms weren't disclosed.   
3913  Trinity said it plans to begin delivery in the...   

                                                      1  
0     [NNP, NNP, CD, NNS, JJ, MD, VB, DT, NN, IN, DT...  
1     [NNP, NNP, VBZ, NN, IN, NNP, NNP, DT, NNP, VBG...  
2     [NNP, NNP, CD, NNS, JJ, CC, JJ, NN, IN, NNP, N...  
3     [DT, NN, IN, NN, RB, VBN, TO, VB, NNP, NN, NNS...  
4     [DT, NN, NN, NN, VBZ, RB, JJ, IN, PRP, VBZ, DT...  
...                                                 ...  
3909  [PRP, RB, VBD, IN, JJR, IN, DT, NN, NNS, VBP, ...  
3910  [NNP, NNP, VBD, IN, DT, JJ, NN, IN, PRP, VBZ, ...  
3911  [NNP, NNPS, NNP, VBD, PRP, VBD, DT, JJ, NN, TO...  
3912                                    [NNS, VBD, VBN]  
3913  [NNP, VBD, PRP, VBZ, TO, VB, NN, IN, DT, JJ, N...  

[3914 rows x 2 columns]

## Checking Unfiltered Data

In [193]:
print(unfl_data[0][0])
print(unfl_data[1][0])

Pierre Vinken, 61 years old, will join the board as a nonexecutive director Nov. 29.
['NNP', 'NNP', 'CD', 'NNS', 'JJ', 'MD', 'VB', 'DT', 'NN', 'IN', 'DT', 'JJ', 'NN', 'NNP', 'CD']


In [194]:
for i in unfl_data[0]:
  print(i)

Pierre Vinken, 61 years old, will join the board as a nonexecutive director Nov. 29.
Mr. Vinken is chairman of Elsevier N.V., the Dutch publishing group.
Rudolph Agnew, 55 years old and former chairman of Consolidated Gold Fields PLC, was named a nonexecutive director of this British industrial conglomerate.
A form of asbestos once used to make Kent cigarette filters has caused a high percentage of cancer deaths among a group of workers exposed to it more than 30 years ago, researchers reported.
The asbestos fiber, crocidolite, is unusually resilient once it enters the lungs, with even brief exposures to it causing symptoms that show up decades later, researchers said.
Lorillard Inc., the unit of New York-based Loews Corp. that makes Kent cigarettes, stopped using crocidolite in its Micronite cigarette filters in 1956.
Although preliminary findings were reported more than a year ago, the latest results appear in today's New England Journal of Medicine, a forum likely to bring new atten

In [195]:
for i in unfl_data[1]:
  print(i)

['NNP', 'NNP', 'CD', 'NNS', 'JJ', 'MD', 'VB', 'DT', 'NN', 'IN', 'DT', 'JJ', 'NN', 'NNP', 'CD']
['NNP', 'NNP', 'VBZ', 'NN', 'IN', 'NNP', 'NNP', 'DT', 'NNP', 'VBG', 'NN']
['NNP', 'NNP', 'CD', 'NNS', 'JJ', 'CC', 'JJ', 'NN', 'IN', 'NNP', 'NNP', 'NNP', 'NNP', 'VBD', 'VBN', 'DT', 'JJ', 'NN', 'IN', 'DT', 'JJ', 'JJ', 'NN']
['DT', 'NN', 'IN', 'NN', 'RB', 'VBN', 'TO', 'VB', 'NNP', 'NN', 'NNS', 'VBZ', 'VBN', 'DT', 'JJ', 'NN', 'IN', 'NN', 'NNS', 'IN', 'DT', 'NN', 'IN', 'NNS', 'VBN', 'TO', 'PRP', 'RBR', 'IN', 'CD', 'NNS', 'IN', 'NNS', 'VBD']
['DT', 'NN', 'NN', 'NN', 'VBZ', 'RB', 'JJ', 'IN', 'PRP', 'VBZ', 'DT', 'NNS', 'IN', 'RB', 'JJ', 'NNS', 'TO', 'PRP', 'VBG', 'NNS', 'WDT', 'VBP', 'RP', 'NNS', 'JJ', 'NNS', 'VBD']
['NNP', 'NNP', 'DT', 'NN', 'IN', 'JJ', 'JJ', 'NNP', 'NNP', 'WDT', 'VBZ', 'NNP', 'NNS', 'VBD', 'VBG', 'NN', 'IN', 'PRP$', 'NN', 'NN', 'NNS', 'IN', 'CD']
['IN', 'JJ', 'NNS', 'VBD', 'VBN', 'RBR', 'IN', 'DT', 'NN', 'IN', 'DT', 'JJS', 'NNS', 'VBP', 'IN', 'NN', 'NNP', 'NNP', 'NNP', 'IN', 'NNP',

## Data Cleaning

In [196]:
# Removing '.' , ',', '`', "'"
# stopwords = ['.',',','`',"'"]

fl_data = []
for i in unfl_data[0]:
  i = i.replace(",","")
  i = i.replace(".","")
  i = i.replace("`","")
  i = i.replace("'","")
  fl_data.append(i)

fl_label = unfl_data[1]

In [197]:
for i in range(10):
  print(fl_data[i])
  print(fl_label[i])

Pierre Vinken 61 years old will join the board as a nonexecutive director Nov 29
['NNP', 'NNP', 'CD', 'NNS', 'JJ', 'MD', 'VB', 'DT', 'NN', 'IN', 'DT', 'JJ', 'NN', 'NNP', 'CD']
Mr Vinken is chairman of Elsevier NV the Dutch publishing group
['NNP', 'NNP', 'VBZ', 'NN', 'IN', 'NNP', 'NNP', 'DT', 'NNP', 'VBG', 'NN']
Rudolph Agnew 55 years old and former chairman of Consolidated Gold Fields PLC was named a nonexecutive director of this British industrial conglomerate
['NNP', 'NNP', 'CD', 'NNS', 'JJ', 'CC', 'JJ', 'NN', 'IN', 'NNP', 'NNP', 'NNP', 'NNP', 'VBD', 'VBN', 'DT', 'JJ', 'NN', 'IN', 'DT', 'JJ', 'JJ', 'NN']
A form of asbestos once used to make Kent cigarette filters has caused a high percentage of cancer deaths among a group of workers exposed to it more than 30 years ago researchers reported
['DT', 'NN', 'IN', 'NN', 'RB', 'VBN', 'TO', 'VB', 'NNP', 'NN', 'NNS', 'VBZ', 'VBN', 'DT', 'JJ', 'NN', 'IN', 'NN', 'NNS', 'IN', 'DT', 'NN', 'IN', 'NNS', 'VBN', 'TO', 'PRP', 'RBR', 'IN', 'CD', 'NNS'

Dataset is now ready in fl_data(_list containing strings_) and fl_label(_list containing list of PoS tags_)

## Train Test Split

In [198]:
from sklearn.model_selection import train_test_split
# X stands for Sentences and Y for the PoS tage associated with them
X = fl_data
Y = fl_label

X_train, X_test, Y_train, Y_test = train_test_split(X,Y,test_size=0.20,shuffle=False)


Now, untill testing do not touch X_test and Y_test

In [199]:
print(X_train)

['Pierre Vinken 61 years old will join the board as a nonexecutive director Nov 29', 'Mr Vinken is chairman of Elsevier NV the Dutch publishing group', 'Rudolph Agnew 55 years old and former chairman of Consolidated Gold Fields PLC was named a nonexecutive director of this British industrial conglomerate', 'A form of asbestos once used to make Kent cigarette filters has caused a high percentage of cancer deaths among a group of workers exposed to it more than 30 years ago researchers reported', 'The asbestos fiber crocidolite is unusually resilient once it enters the lungs with even brief exposures to it causing symptoms that show up decades later researchers said', 'Lorillard Inc the unit of New York-based Loews Corp that makes Kent cigarettes stopped using crocidolite in its Micronite cigarette filters in 1956', 'Although preliminary findings were reported more than a year ago the latest results appear in todays New England Journal of Medicine a forum likely to bring new attention to

In [200]:
print(len(X_train))
print(len(Y_train))

3131
3131


So the original dataset was of length 3914, It has been divided into 3131 as Training Set and 783 as Test Set


In [201]:
for i in range(10):
  print(X_train[i]," ",len(X_train[i].split(" ")))
  print(Y_train[i]," ",len(Y_train[i]))

Pierre Vinken 61 years old will join the board as a nonexecutive director Nov 29   15
['NNP', 'NNP', 'CD', 'NNS', 'JJ', 'MD', 'VB', 'DT', 'NN', 'IN', 'DT', 'JJ', 'NN', 'NNP', 'CD']   15
Mr Vinken is chairman of Elsevier NV the Dutch publishing group   11
['NNP', 'NNP', 'VBZ', 'NN', 'IN', 'NNP', 'NNP', 'DT', 'NNP', 'VBG', 'NN']   11
Rudolph Agnew 55 years old and former chairman of Consolidated Gold Fields PLC was named a nonexecutive director of this British industrial conglomerate   23
['NNP', 'NNP', 'CD', 'NNS', 'JJ', 'CC', 'JJ', 'NN', 'IN', 'NNP', 'NNP', 'NNP', 'NNP', 'VBD', 'VBN', 'DT', 'JJ', 'NN', 'IN', 'DT', 'JJ', 'JJ', 'NN']   23
A form of asbestos once used to make Kent cigarette filters has caused a high percentage of cancer deaths among a group of workers exposed to it more than 30 years ago researchers reported   34
['DT', 'NN', 'IN', 'NN', 'RB', 'VBN', 'TO', 'VB', 'NNP', 'NN', 'NNS', 'VBZ', 'VBN', 'DT', 'JJ', 'NN', 'IN', 'NN', 'NNS', 'IN', 'DT', 'NN', 'IN', 'NNS', 'VBN', 'T

Everything Looks Good!

# 36 Tag System

## Calculation of Probability Matrices

In [202]:
name_set = set()

In [203]:
for i in X_train:
  tokens = i.split(' ')
  for j in tokens:
    name_set.add(j)

In [204]:
len(name_set)

11011

In [205]:
tag_set = set()

for i in Y_train:
  for j in i:
    tag_set.add(j)

In [206]:
tag_set.remove('#')
tag_set.remove("''")
tag_set.remove(",")
tag_set.remove(":")
tag_set.remove("-LRB-")
tag_set.remove("-RRB-")

In [207]:
tag_set

{'CC',
 'CD',
 'DT',
 'EX',
 'FW',
 'IN',
 'JJ',
 'JJR',
 'JJS',
 'LS',
 'MD',
 'NN',
 'NNP',
 'NNPS',
 'NNS',
 'PDT',
 'PRP',
 'PRP$',
 'RB',
 'RBR',
 'RBS',
 'RP',
 'SYM',
 'TO',
 'UH',
 'VB',
 'VBD',
 'VBG',
 'VBN',
 'VBP',
 'VBZ',
 'WDT',
 'WP',
 'WP$',
 'WRB'}

In [208]:
len(tag_set)

35

Looks Fine!!

In [209]:
df = pd.DataFrame()


df['Names'] = list(name_set)


for i in tag_set:
  df[i] = [1.0 for i in range(1,11012)]

df = df.drop(index = 0)
df.set_index("Names", inplace = True)

df.head(10)

TO  VBG   NN  JJS  NNS  PRP   CD   JJ   EX  WP$  ...   WP   LS  \
Names                                                           ...             
talk          1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0   
853%          1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0   
collections   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0   
1901          1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0   
psychiatrist  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0   
Al            1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0   
laboriously   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0   
twindam       1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0   
Cellars       1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0   
812%          1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0   

               UH   CC   RB  VBN  PRP$  WRB  NNP   RP  
Names                                                  
talk          1.0  1.0  1.0  1.0   1.0  1.0  1.0  1.0  
853%          1.0  1.0  1.0  1.0   1.0  1.0  1.0  1.0  
collections   1.0  1.0  1.0  1.0   1.0  1.0  1.0  1.0  
1901          1.0  1.0  1.0  1.0   1.0  1.0  1.0  1.0  
psychiatrist  1.0  1.0  1.0  1.0   1.0  1.0  1.0  1.0  
Al            1.0  1.0  1.0  1.0   1.0  1.0  1.0  1.0  
laboriously   1.0  1.0  1.0  1.0   1.0  1.0  1.0  1.0  
twindam       1.0  1.0  1.0  1.0   1.0  1.0  1.0  1.0  
Cellars       1.0  1.0  1.0  1.0   1.0  1.0  1.0  1.0  
812%          1.0  1.0  1.0  1.0   1.0  1.0  1.0  1.0  

[10 rows x 35 columns]

In [210]:
df.shape

(11010, 35)

In [211]:
df.loc['coaching','PRP']

1.0

In [212]:
# Simply traverse all the training set and map and increase the values appropriately in the matrix
for i in range(3131):
  words = X_train[i].split(' ');
  tags = Y_train[i]
  for j in range(len(words)):
    w = words[j];
    t = tags[j];
    # print(w,"->",t)\
    try:
      df.loc[w, t] = df.loc[w,t] + 1
    except:
      print("Exception occured for ",w," ",t);

Exception occured for  --   :
Exception occured for  --   :
Exception occured for  --   :
Exception occured for  --   :
Exception occured for  --   :
Exception occured for  --   :
Exception occured for  --   :
Exception occured for  --   :
Exception occured for     :
Exception occured for  --   :
Exception occured for  --   :
Exception occured for     :
Exception occured for  --   :
Exception occured for  --   :
Exception occured for  --   :
Exception occured for  --   :
Exception occured for  --   :
Exception occured for  --   :
Exception occured for  --   :
Exception occured for  --   :
Exception occured for  -LCB-   -LRB-
Exception occured for  -RCB-   -RRB-
Exception occured for  --   :
Exception occured for  --   :
Exception occured for  --   :
Exception occured for  -LRB-   -LRB-
Exception occured for  -RRB-   -RRB-
Exception occured for  --   :
Exception occured for  --   :
Exception occured for  -LRB-   -LRB-
Exception occured for  -RRB-   -RRB-
Exception occured for  --   :
Ex

Found Some Outliers and neglecting them

In [213]:
df.head

<bound method NDFrame.head of                TO  VBG   NN  JJS  NNS  PRP   CD   JJ   EX  WP$  ...   WP   LS  \
Names                                                           ...             
talk          1.0  1.0  3.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0   
853%          1.0  1.0  1.0  1.0  1.0  1.0  2.0  1.0  1.0  1.0  ...  1.0  1.0   
collections   1.0  1.0  1.0  1.0  3.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0   
1901          1.0  1.0  1.0  1.0  1.0  1.0  2.0  1.0  1.0  1.0  ...  1.0  1.0   
psychiatrist  1.0  1.0  2.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0   
...           ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
circulated    1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0   
payoff        1.0  1.0  2.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0   
flap          1.0  1.0  3.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0   
Mifflin       1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  

In [214]:
# Finding sum of each column and then dividing at each column
sum_tags = []
for i in tag_set:
  sum = df[i].sum()
  sum_tags.append(sum)

In [215]:
sum_tags

[12725.0,
 12209.0,
 20947.0,
 11154.0,
 15930.0,
 12504.0,
 13524.0,
 15756.0,
 11091.0,
 11020.0,
 11190.0,
 13310.0,
 11351.0,
 11014.0,
 11040.0,
 11011.0,
 11728.0,
 11315.0,
 17564.0,
 18915.0,
 12715.0,
 11130.0,
 11031.0,
 12126.0,
 13066.0,
 11225.0,
 11023.0,
 11013.0,
 12846.0,
 13080.0,
 12692.0,
 11654.0,
 11161.0,
 18620.0,
 11192.0]

In [216]:

for j in range(35):
  for i in range(11010):
    df.iloc[i][j] = float(df.iloc[i][j]/sum_tags[j])


In [217]:
print(df.head)

<bound method NDFrame.head of                     TO       VBG        NN      JJS       NNS      PRP  \
Names                                                                    
talk          0.000079  0.000082  0.000143  0.00009  0.000063  0.00008   
853%          0.000079  0.000082  0.000048  0.00009  0.000063  0.00008   
collections   0.000079  0.000082  0.000048  0.00009  0.000188  0.00008   
1901          0.000079  0.000082  0.000048  0.00009  0.000063  0.00008   
psychiatrist  0.000079  0.000082  0.000095  0.00009  0.000063  0.00008   
...                ...       ...       ...      ...       ...      ...   
circulated    0.000079  0.000082  0.000048  0.00009  0.000063  0.00008   
payoff        0.000079  0.000082  0.000095  0.00009  0.000063  0.00008   
flap          0.000079  0.000082  0.000143  0.00009  0.000063  0.00008   
Mifflin       0.000079  0.000082  0.000048  0.00009  0.000063  0.00008   
economists    0.000079  0.000082  0.000048  0.00009  0.000314  0.00008   

       

Emission Matrix is ready
Lets create Transmission Matrix Too

In [218]:
df_tp = pd.DataFrame()

tags_row = list(tag_set)
tags_row.insert(0,'S')

tags_cols = list(tag_set)
last = len(tag_set)
tags_cols.insert(last,'E')


# print(tags_row)
# print(tags_cols)
df_tp['Tags'] = tags_row


for i in tags_cols:
  df_tp[i] = [1.0 for i in range(len(tags_cols))]

df_tp.set_index("Tags", inplace = True)

df_tp.shape
df_tp

TO  VBG   NN  JJS  NNS  PRP   CD   JJ   EX  WP$  ...   LS   UH   CC  \
Tags                                                    ...                  
S     1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
TO    1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
VBG   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
NN    1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
JJS   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
NNS   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
PRP   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
CD    1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
JJ    1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
EX    1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
WP$   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
NNPS  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
VBD   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
WDT   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
FW    1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
RBS   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
SYM   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
MD    1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
JJR   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
DT    1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
IN    1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
VBZ   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
RBR   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
PDT   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
VBP   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
VB    1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
WP    1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
LS    1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
UH    1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
CC    1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
RB    1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
VBN   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
PRP$  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
WRB   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
NNP   1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   
RP    1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  1.0  1.0   

       RB  VBN  PRP$  WRB  NNP   RP    E  
Tags                                      
S     1.0  1.0   1.0  1.0  1.0  1.0  1.0  
TO    1.0  1.0   1.0  1.0  1.0  1.0  1.0  
VBG   1.0  1.0   1.0  1.0  1.0  1.0  1.0  
NN    1.0  1.0   1.0  1.0  1.0  1.0  1.0  
JJS   1.0  1.0   1.0  1.0  1.0  1.0  1.0  
NNS   1.0  1.0   1.0  1.0  1.0  1.0  1.0  
PRP   1.0  1.0   1.0  1.0  1.0  1.0  1.0  
CD    1.0  1.0   1.0  1.0  1.0  1.0  1.0  
JJ    1.0  1.0   1.0  1.0  1.0  1.0  1.0  
EX    1.0  1.0   1.0  1.0  1.0  1.0  1.0  
WP$   1.0  1.0   1.0  1.0  1.0  1.0  1.0  
NNPS  1.0  1.0   1.0  1.0  1.0  1.0  1.0  
VBD   1.0  1.0   1.0  1.0  1.0  1.0  1.0  
WDT   1.0  1.0   1.0  1.0  1.0  1.0  1.0  
FW    1.0  1.0   1.0  1.0  1.0  1.0  1.0  
RBS   1.0  1.0   1.0  1.0  1.0  1.0  1.0  
SYM   1.0  1.0   1.0  1.0  1.0  1.0  1.0  
MD    1.0  1.0   1.0  1.0  1.0  1.0  1.0  
JJR   1.0  1.0   1.0  1.0  1.0  1.0  1.0  
DT    1.0  1.0   1.0  1.0  1.0  1.0  1.0  
IN    1.0  1.0   1.0  1.0  1.0  1.0  1.0  
VBZ   1.0  1.0   1.0  1.0  1.0  1.0  1.0  
RBR   1.0 

In [219]:
# Setting up S and E

for i in Y_train:
  start_tag = i[0]
  last_index = len(i)-1
  last_tag = i[last_index]
  try:
    df_tp.loc['S'][start_tag] = df_tp.loc['S'][start_tag] + 1
    df_tp.loc[last_tag]['E'] = df_tp.loc[last_tag]['E'] + 1
  except:
    print("Exception Occured for ", start_tag)
    print("Exception Occured for ", last_tag)

df_tp.head(36) 



Exception Occured for  DT
Exception Occured for  :
Exception Occured for  -LRB-
Exception Occured for  -RRB-
Exception Occured for  NNP
Exception Occured for  -RRB-
Exception Occured for  NNP
Exception Occured for  :
Exception Occured for  RB
Exception Occured for  -RRB-
Exception Occured for  -LRB-
Exception Occured for  -RRB-
Exception Occured for  NNP
Exception Occured for  -RRB-
Exception Occured for  DT
Exception Occured for  :
Exception Occured for  IN
Exception Occured for  :
Exception Occured for  PRP
Exception Occured for  ''
Exception Occured for  NNP
Exception Occured for  -RRB-
Exception Occured for  NNP
Exception Occured for  :
Exception Occured for  -LRB-
Exception Occured for  -RRB-
Exception Occured for  NNP
Exception Occured for  :
Exception Occured for  -LRB-
Exception Occured for  -RRB-
Exception Occured for  IN
Exception Occured for  -RRB-
Exception Occured for  NNP
Exception Occured for  -RRB-
Exception Occured for  VBG
Exception Occured for  -RRB-
Exception Occure

TO   VBG     NN  JJS    NNS    PRP    CD     JJ    EX  WP$  ...   LS  \
Tags                                                                ...        
S     17.0  39.0  132.0  5.0  145.0  308.0  33.0  127.0  30.0  1.0  ...  8.0   
TO     1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
VBG    1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
NN     1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
JJS    1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
NNS    1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
PRP    1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
CD     1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
JJ     1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
EX     1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
WP$    1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
NNPS   1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
VBD    1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
WDT    1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
FW     1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
RBS    1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
SYM    1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
MD     1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
JJR    1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
DT     1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
IN     1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
VBZ    1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
RBR    1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
PDT    1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
VBP    1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
VB     1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
WP     1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
LS     1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
UH     1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
CC     1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
RB     1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
VBN    1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
PRP$   1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
WRB    1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
NNP    1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   
RP     1.0   1.0    1.0  1.0    1.0    1.0   1.0    1.0   1.0  1.0  ...  1.0   

       UH     CC     RB   VBN  PRP$   WRB    NNP   RP       E  
Tags                                                           
S     2.0  177.0  170.0  24.0  26.0  27.0  574.0  1.0     1.0  
TO    1.0    1.0    1.0   1.0   1.0   1.0    1.0  1.0     2.0  
VBG   1.0    1.0    1.0   1.0   1.0   1.0    1.0  1.0    25.0  
NN    1.0    1.0    1.0   1.0   1.0   1.0    1.0  1.0  1062.0  
JJS   1.0    1.0    1.0   1.0   1.0   1.0    1.0  1.0     5.0  
NNS   1.0    1.0    1.0   1.0   1.0   1.0    1.0  1.0   633.0  
PRP   1.0    1.0    1.0   1.0   1.0   1.0    1.0  1.0    48.0  
CD    1.0    1.0    1.0   1.0   1.0   1.0    1.0  1.0   240.0  
JJ    1.0    1.0    1.0   1.0   1.0   1.0    1.0  1.0   107.0  
EX    1.0    1.0    1.0   1.0   1.0   1.0    1.0  1.0     1.0  
WP$   1.0    1.0    1.0   1.0   1.0   1.0    1.0  1.0     1.0  
NNPS  1.0    1.0    1.0   1.0   1.0   1.0    1.0  1.0    22.0  
VBD   1.0    1.0    1.0   1.0   1.0   1.0    1.0  1.0   119.0  
WDT   1

In [220]:
for i in Y_train:
  curr = i
  for j in range(1,len(curr)):
    col = i[j]
    row = i[j-1]
    try:
      df_tp.loc[row][col] = df_tp.loc[row][col] + 1
    except:
      print("Exception Occured!")

Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Occured!
Exception Oc

In [221]:
tp_sum = df_tp.sum(axis=1)

In [222]:
df_tp

TO    VBG      NN   JJS     NNS    PRP     CD      JJ    EX  WP$  \
Tags                                                                        
S      17.0   39.0   132.0   5.0   145.0  308.0   33.0   127.0  30.0  1.0   
TO      1.0   11.0    45.0   1.0    55.0   12.0  149.0    53.0   1.0  1.0   
VBG   107.0    6.0   170.0   1.0   115.0   44.0   32.0    92.0   1.0  1.0   
NN    377.0  138.0  1430.0   9.0   891.0  140.0   50.0   169.0   9.0  2.0   
JJS     1.0    5.0    39.0   1.0    25.0    2.0    9.0    21.0   1.0  1.0   
NNS   198.0  113.0   156.0   1.0   114.0   60.0   46.0   117.0   5.0  5.0   
PRP    32.0   26.0    11.0   1.0     4.0   11.0    2.0    30.0   1.0  1.0   
CD    121.0   22.0   273.0   5.0   392.0   14.0  487.0   120.0   1.0  2.0   
JJ    121.0   28.0  2090.0   3.0  1220.0   22.0  109.0   363.0   1.0  1.0   
EX      1.0    1.0     3.0   1.0     1.0    1.0    1.0     1.0   1.0  1.0   
WP$     1.0    1.0     7.0   1.0     4.0    1.0    1.0     2.0   1.0  1.0   
NNPS    3.0    3.0    11.0   1.0     4.0    1.0    2.0     2.0   1.0  1.0   
VBD   143.0   59.0    94.0   2.0    59.0  159.0  116.0   115.0   5.0  1.0   
WDT     2.0    1.0     4.0   2.0     7.0   15.0    3.0     7.0   2.0  1.0   
FW      1.0    1.0     2.0   1.0     1.0    1.0    1.0     1.0   1.0  1.0   
RBS     1.0    1.0     1.0   1.0     1.0    1.0    1.0    22.0   1.0  1.0   
SYM     1.0    1.0     1.0   1.0     1.0    1.0    1.0     1.0   1.0  1.0   
MD      4.0    2.0     1.0   1.0     2.0    2.0    1.0     2.0   1.0  1.0   
JJR     9.0    1.0    62.0   1.0    66.0    1.0    3.0    23.0   2.0  1.0   
DT      4.0   77.0  3064.0  63.0   559.0    9.0  197.0  1359.0   2.0  1.0   
IN     19.0  253.0   862.0  37.0   523.0  289.0  633.0   787.0  12.0  1.0   
VBZ    98.0   93.0    68.0   3.0    26.0   58.0   42.0   122.0   3.0  1.0   
RBR     3.0    3.0     1.0   1.0     1.0    2.0    1.0    47.0   1.0  1.0   
PDT     1.0    1.0     1.0   1.0     1.0    1.0    1.0     1.0   1.0  1.0   
VBP    70.0   86.0    38.0   1.0    26.0   50.0   12.0    98.0   2.0  1.0   
VB     83.0   32.0   133.0   2.0    98.0   92.0   65.0   185.0   2.0  2.0   
WP      2.0    3.0     4.0   1.0     3.0   13.0    2.0     5.0   1.0  1.0   
LS      1.0    1.0     1.0   1.0     1.0    1.0    1.0     1.0   1.0  1.0   
UH      1.0    1.0     1.0   1.0     1.0    2.0    1.0     1.0   1.0  1.0   
CC     16.0   50.0   211.0   6.0   132.0   90.0   86.0   213.0  12.0  1.0   
RB     55.0   85.0    42.0   2.0    28.0   33.0   82.0   216.0   6.0  1.0   
VBN   181.0   35.0   107.0   1.0    71.0   22.0   50.0    98.0   2.0  1.0   
PRP$    1.0    5.0   277.0   9.0   142.0    1.0   19.0   150.0   1.0  1.0   
WRB     3.0    2.0     9.0   1.0    12.0   28.0    4.0    13.0   2.0  1.0   
NNP    55.0   37.0   595.0  10.0   209.0   28.0  252.0   176.0   1.0  3.0   
RP     10.0    3.0    10.0   1.0    10.0    3.0    6.0    12.0   1.0  1.0   

      ...   LS   UH     CC     RB    VBN   PRP$   WRB     NNP    RP       E  
Tags  ...                                                                    
S     ...  8.0  2.0  177.0  170.0   24.0   26.0  27.0   574.0   1.0     1.0  
TO    ...  1.0  1.0    2.0   15.0    4.0   32.0   1.0    71.0   1.0     2.0  
VBG   ...  1.0  1.0   23.0   51.0   27.0   36.0   1.0    35.0  25.0    25.0  
NN    ...  2.0  1.0  492.0  245.0  117.0   11.0  32.0   282.0   6.0  1062.0  
JJS   ...  1.0  1.0    2.0    6.0    4.0    1.0   1.0     3.0   1.0     5.0  
NNS   ...  2.0  1.0  365.0  185.0  120.0    7.0  14.0    49.0   2.0   633.0  
PRP   ...  1.0  1.0   24.0  108.0   13.0    1.0   1.0     8.0   7.0    48.0  
CD    ...  1.0  1.0  104.0   29.0   28.0    1.0   7.0    90.0   1.0   240.0  
JJ    ...  1.0  1.0   98.0   28.0   16.0    2.0   5.0   173.0   2.0   107.0  
EX    ...  1.0  1.0    1.0    4.0    1.0    1.0   1.0     1.0   1.0     1.0  
WP$   ...  1.0  1.0    1.0    1.0    1.0    1.0   1.0     1.0   1.0     1.0  
NNPS  ...  1.0  1.0   13.0    3.0    3.0    2.0   1.0    66.0   1.0 

In [223]:
tp_sum = list(tp_sum)

tp_sum

[3151.0,
 1745.0,
 1233.0,
 9864.0,
 179.0,
 4870.0,
 1522.0,
 2512.0,
 4766.0,
 117.0,
 46.0,
 211.0,
 2333.0,
 377.0,
 39.0,
 66.0,
 37.0,
 754.0,
 340.0,
 6586.0,
 7933.0,
 1737.0,
 155.0,
 57.0,
 1151.0,
 2083.0,
 251.0,
 41.0,
 39.0,
 1871.0,
 2095.0,
 1711.0,
 679.0,
 187.0,
 7552.0,
 218.0]

In [224]:
len(tp_sum)

36

In [225]:
for i in range(36):
  for j in range(36):
    df_tp.iloc[i][j] = df_tp.iloc[i][j]/tp_sum[i]

df_tp

TO       VBG        NN       JJS       NNS       PRP        CD  \
Tags                                                                         
S     0.005395  0.012377  0.041891  0.001587  0.046017  0.097747  0.010473   
TO    0.000573  0.006304  0.025788  0.000573  0.031519  0.006877  0.085387   
VBG   0.086780  0.004866  0.137875  0.000811  0.093268  0.035685  0.025953   
NN    0.038220  0.013990  0.144972  0.000912  0.090328  0.014193  0.005069   
JJS   0.005587  0.027933  0.217877  0.005587  0.139665  0.011173  0.050279   
NNS   0.040657  0.023203  0.032033  0.000205  0.023409  0.012320  0.009446   
PRP   0.021025  0.017083  0.007227  0.000657  0.002628  0.007227  0.001314   
CD    0.048169  0.008758  0.108678  0.001990  0.156051  0.005573  0.193869   
JJ    0.025388  0.005875  0.438523  0.000629  0.255980  0.004616  0.022870   
EX    0.008547  0.008547  0.025641  0.008547  0.008547  0.008547  0.008547   
WP$   0.021739  0.021739  0.152174  0.021739  0.086957  0.021739  0.021739   
NNPS  0.014218  0.014218  0.052133  0.004739  0.018957  0.004739  0.009479   
VBD   0.061294  0.025289  0.040291  0.000857  0.025289  0.068153  0.049721   
WDT   0.005305  0.002653  0.010610  0.005305  0.018568  0.039788  0.007958   
FW    0.025641  0.025641  0.051282  0.025641  0.025641  0.025641  0.025641   
RBS   0.015152  0.015152  0.015152  0.015152  0.015152  0.015152  0.015152   
SYM   0.027027  0.027027  0.027027  0.027027  0.027027  0.027027  0.027027   
MD    0.005305  0.002653  0.001326  0.001326  0.002653  0.002653  0.001326   
JJR   0.026471  0.002941  0.182353  0.002941  0.194118  0.002941  0.008824   
DT    0.000607  0.011691  0.465229  0.009566  0.084877  0.001367  0.029912   
IN    0.002395  0.031892  0.108660  0.004664  0.065927  0.036430  0.079793   
VBZ   0.056419  0.053541  0.039148  0.001727  0.014968  0.033391  0.024180   
RBR   0.019355  0.019355  0.006452  0.006452  0.006452  0.012903  0.006452   
PDT   0.017544  0.017544  0.017544  0.017544  0.017544  0.017544  0.017544   
VBP   0.060817  0.074718  0.033015  0.000869  0.022589  0.043440  0.010426   
VB    0.039846  0.015362  0.063850  0.000960  0.047048  0.044167  0.031205   
WP    0.007968  0.011952  0.015936  0.003984  0.011952  0.051793  0.007968   
LS    0.024390  0.024390  0.024390  0.024390  0.024390  0.024390  0.024390   
UH    0.025641  0.025641  0.025641  0.025641  0.025641  0.051282  0.025641   
CC    0.008552  0.026724  0.112774  0.003207  0.070551  0.048103  0.045965   
RB    0.026253  0.040573  0.020048  0.000955  0.013365  0.015752  0.039141   
VBN   0.105786  0.020456  0.062537  0.000584  0.041496  0.012858  0.029223   
PRP$  0.001473  0.007364  0.407953  0.013255  0.209131  0.001473  0.027982   
WRB   0.016043  0.010695  0.048128  0.005348  0.064171  0.149733  0.021390   
NNP   0.007283  0.004899  0.078787  0.001324  0.027675  0.003708  0.033369   
RP    0.045872  0.013761  0.045872  0.004587  0.045872  0.013761  0.027523   

            JJ        EX       WP$  ...        LS        UH        CC  \
Tags                                ...                                 
S     0.040305  0.009521  0.000317  ...  0.002539  0.000635  0.056173   
TO    0.030372  0.000573  0.000573  ...  0.000573  0.000573  0.001146   
VBG   0.074615  0.000811  0.000811  ...  0.000811  0.000811  0.018654   
NN    0.017133  0.000912  0.000203  ...  0.000203  0.000101  0.049878   
JJS   0.117318  0.005587  0.005587  ...  0.005587  0.005587  0.011173   
NNS   0.024025  0.001027  0.001027  ...  0.000411  0.000205  0.074949   
PRP   0.019711  0.000657  0.000657  ...  0.000657  0.000657  0.015769   
CD    0.047771  0.000398  0.000796  ...  0.000398  0.000398  0.041401   
JJ    0.076164  0.000210  0.000210  ...  0.000210  0.000210  0.020562   
EX    0.008547  0.008547  0.008547  ...  0.008547  0.008547  0.008547   
WP$   0.043478  0.021739  0.021739  ...  0.021739  0.021739  0.021739   
NNPS  0.009479  0.004739  0.004739  ...  0.004739  0.004739  0.061611   
VBD   0.049293  0.002143 

In [226]:
name = df_tp['CD'].idxmax()
print(name)

CD


>So now both the matrices are ready

>Let's calculate through viterbi algorithm

In [227]:
def viterbi_algo(sentence):
  sentence = sentence.replace(",","")
  sentence = sentence.replace(".","")
  sentence = sentence.replace("`","")
  sentence = sentence.replace("'","")
  sentence = sentence.replace("?","")
  words = sentence.split(" ")

  df_words = pd.DataFrame()
  tag_list = list(tag_set)
  df_words['Tags'] = tag_list
  res_tag_list = []
  for  i in words:
    df_words[i] = [1.0 for i in range(len(tag_set))]

  df_words.set_index("Tags", inplace = True)

  flag = False
  start_word = words[0]
  if(start_word not in name_set):
    flag = True
  for i in range(len(tag_set)):
    if flag == False:
      em_prob = df.loc[start_word][tag_list[i]]
    else:
      em_prob = 1
    t_prob = df_tp.loc['S'][tag_list[i]]
    res_prob = em_prob * t_prob

    # print("Prob for ",start_word,"and Tag: ",tag_list[i],"EmP:",em_prob," TrP:",t_prob," Res: ",res_prob)
    df_words.loc[tag_list[i]][start_word] = df_words.loc[tag_list[i]][start_word] * res_prob
  
  for i in range(1,len(words)):
    flag = False
    # find maximum prob of prev col and tag associated with it. Add the tag into result list
    curr_word = words[i]
    if(curr_word not in name_set):
      flag = True
    prev_col = df_words.columns[i-1]
    last_res_prob = df_words[prev_col].max()
    last_res_tag = df_words[prev_col].idxmax()
    res_tag_list.append(last_res_tag)

    for j in range(len(tag_list)):
      if(flag == False):
        curr_em_prob = df.loc[curr_word][tag_list[j]] 
      else:
        curr_em_prob = 1    
      curr_t_prob = df_tp.loc[last_res_tag][tag_list[j]]
      res_prob = curr_em_prob * curr_t_prob * last_res_prob
      df_words.loc[tag_list[j]][curr_word] = df_words.loc[tag_list[j]][curr_word] * res_prob  

  # print(df_words)
  return [res_prob,words,res_tag_list]

In [228]:
res_prob,words,res_tag_list = viterbi_algo("Sales rose 10% to $13.8 billion from $12.5 billion.")

In [229]:
words

['Sales', 'rose', '10%', 'to', '$138', 'billion', 'from', '$125', 'billion']

In [230]:
res_tag_list

['NNS', 'VBD', 'CD', 'TO', 'VB', 'CD', 'IN', 'DT']

> Code is looking to be working fine!

> Let us test this model now

In [231]:
X_test[16]

'The Ginnie Mae 9% securities were yielding 932% to a 12-year average life'

In [232]:
Y_test

3131    [DT, JJS, JJ, NNS, VBD, VBN, IN, CD, CD, TO, V...
3132    [DT, NN, NN, IN, JJ, NNP, NNS, VBD, RB, JJ, IN...
3133                                          [NNP, NNPS]
3134    [NNP, CD, CD, NN, NN, VBD, NN, IN, DT, JJ, NN,...
3135    [RB, RBS, JJ, NNS, VBD, JJ, TO, RB, JJ, IN, CD...
                              ...                        
3909    [PRP, RB, VBD, IN, JJR, IN, DT, NN, NNS, VBP, ...
3910    [NNP, NNP, VBD, IN, DT, JJ, NN, IN, PRP, VBZ, ...
3911    [NNP, NNPS, NNP, VBD, PRP, VBD, DT, JJ, NN, TO...
3912                                      [NNS, VBD, VBN]
3913    [NNP, VBD, PRP, VBZ, TO, VB, NN, IN, DT, JJ, N...
Name: 1, Length: 783, dtype: object

In [233]:
print(len(X_test))
print(type(Y_test))

783
<class 'pandas.core.series.Series'>


In [234]:
predicted_res_list = []
indices_list = []
for i in range(len(X_test)):
  try:
    res_prob,words,res_tag_list = viterbi_algo(X_test[i])
    indices_list.append(i)
    predicted_res_list.append(res_tag_list)
  except:
    continue
# predicted_res_list

In [235]:
print(len(predicted_res_list))
print(len(indices_list))

490
490


In [236]:
print(indices_list)

[2, 3, 4, 6, 7, 9, 11, 12, 13, 14, 16, 17, 18, 19, 20, 22, 23, 25, 27, 28, 30, 32, 33, 34, 35, 36, 38, 40, 41, 45, 49, 50, 51, 52, 54, 55, 56, 59, 60, 61, 63, 68, 69, 70, 71, 72, 74, 75, 76, 79, 80, 83, 85, 87, 89, 90, 92, 93, 99, 101, 103, 104, 105, 106, 107, 110, 112, 113, 114, 115, 116, 117, 118, 119, 121, 124, 125, 126, 127, 128, 130, 132, 133, 135, 137, 139, 140, 142, 143, 144, 146, 147, 148, 149, 150, 151, 153, 154, 155, 156, 157, 161, 162, 164, 165, 166, 168, 169, 173, 175, 176, 177, 179, 180, 185, 188, 189, 190, 192, 195, 198, 199, 201, 202, 203, 204, 205, 206, 207, 209, 212, 215, 216, 217, 218, 219, 220, 222, 223, 224, 228, 229, 231, 232, 234, 235, 236, 238, 239, 240, 241, 242, 243, 244, 248, 249, 251, 253, 254, 255, 256, 259, 260, 261, 262, 263, 264, 267, 268, 272, 273, 274, 275, 280, 281, 282, 283, 285, 294, 295, 296, 297, 298, 300, 301, 302, 303, 304, 306, 308, 310, 311, 312, 313, 314, 315, 316, 317, 319, 320, 321, 322, 324, 326, 329, 331, 335, 337, 339, 340, 341, 344, 347,

In [237]:
Y_test = Y_test.tolist()

In [238]:
actual_res_list = []
for i in indices_list:
  actual_res_list.append(Y_test[i])

len(actual_res_list)

490

In [239]:
print(predicted_res_list)
print(actual_res_list)

[['NNP'], ['NNP', 'NNP', 'CD', 'CD', 'NN', 'IN', 'NN', 'IN', 'DT', 'JJ', 'NN'], ['RB', 'IN', 'JJ', 'NNS', 'VBD', 'JJ', 'TO', 'IN', 'JJ', 'TO', 'TO', 'TO'], ['NNP', 'VBD', 'DT', 'NNS', 'IN', 'IN', 'DT', 'NN', 'NN', 'NN', 'IN', 'DT', 'NNP', 'NNS', 'CC'], ['DT', 'NN', 'NNS', 'NN', 'NNS', 'CC', 'NN'], ['DT', 'NNS', 'IN', 'DT', 'NN'], ['NNP', 'NNP', 'IN', 'NNS', 'IN', 'NN', 'VBP', 'VBN', 'IN', 'DT', 'NN'], ['NNP', 'VBD', 'DT', 'JJ', 'JJ', 'NN', 'NN', 'IN', 'CD', 'CD', 'JJ', 'NNS', 'VBD', 'IN', 'DT', 'IN'], ['DT'], ['NNP', 'NNS', 'VBD', 'RB', 'JJR', 'CC', 'NNP', 'NNP', 'IN', 'DT', 'NNP'], ['DT', 'NN', 'IN', 'CD', 'NNS', 'VBD', 'DT', 'NN', 'TO', 'DT', 'NN', 'NN'], ['DT', 'VBD', 'JJ', 'IN', 'DT', 'NN', 'IN', 'DT', 'JJ', 'NNS'], ['NNP'], ['NNP', 'NNS', 'VBD', 'RB', 'JJ', 'TO', 'VB', 'DT', 'NN', 'IN', 'JJ', 'NN', 'NN', 'IN', 'TO', 'TO', 'TO'], ['DT', 'NN', 'CD', 'NN', 'IN', 'NN', 'NN', 'NN', 'NNS', 'IN', 'DT', 'NNP', 'NNP', 'NNP', 'NNP', 'NNP', 'VBD', 'RB', 'VBN', 'IN', 'VBG', 'RB', 'VBN', 'IN',

In [240]:
for i in range(490):
  actual_res_list[i].pop(-1)

In [241]:
for i in range(490):
  if((len(predicted_res_list[i])) != (len(actual_res_list[i]))):
    print("Not Equal")
  else:
    print("Equal with length: ",len(predicted_res_list[i]))
  # print(len(predicted_res_list[i]))
  # print(len(actual_res_list[i]))

Equal with length:  1
Equal with length:  11
Equal with length:  12
Equal with length:  15
Equal with length:  7
Equal with length:  5
Equal with length:  11
Equal with length:  16
Equal with length:  1
Equal with length:  11
Equal with length:  12
Equal with length:  10
Equal with length:  1
Equal with length:  17
Equal with length:  28
Equal with length:  19
Equal with length:  40
Equal with length:  22
Equal with length:  1
Equal with length:  17
Equal with length:  12
Equal with length:  0
Equal with length:  20
Equal with length:  17
Equal with length:  12
Equal with length:  16
Equal with length:  17
Equal with length:  8
Equal with length:  19
Equal with length:  19
Equal with length:  8
Equal with length:  18
Equal with length:  16
Equal with length:  29
Equal with length:  16
Equal with length:  20
Equal with length:  17
Equal with length:  21
Equal with length:  18
Equal with length:  13
Equal with length:  13
Equal with length:  21
Equal with length:  29
Equal with length:  

In [242]:
actual_res_series = pd.Series(actual_res_list)
predicted_res_series = pd.Series(predicted_res_list)

In [243]:
from sklearn.metrics import f1_score
from sklearn.metrics import accuracy_score

from sklearn.preprocessing import MultiLabelBinarizer

actual_res = MultiLabelBinarizer().fit_transform(actual_res_series)
predicted_res = MultiLabelBinarizer().fit_transform(predicted_res_series)

In [244]:
print(actual_res_series[1])
print(predicted_res_series[1])

['NNP', 'CD', 'CD', 'NN', 'NN', 'VBD', 'NN', 'IN', 'DT', 'JJ', 'NN']
['NNP', 'NNP', 'CD', 'CD', 'NN', 'IN', 'NN', 'IN', 'DT', 'JJ', 'NN']


In [255]:
count = 0
for i in range(490):
  if(len(actual_res_series[i]) == 0):
    count = count+1
print(count)

3


In [260]:
score = 0
for i in range(490):
  if(len(actual_res_series[i]) != 0):
    score += accuracy_score(actual_res_series[i],predicted_res_series[i])

print(score)

330.3993739762026


In [256]:
f1_av_score=0
acc_av_score=0
for i in range(490):
  f1_av_score += f1_score(actual_res_series[i],predicted_res_series[i],average='micro')
  if(len(actual_res_series[i]) != 0):
    acc_av_score += accuracy_score(actual_res_series[i],predicted_res_series[i])

tag_36_f1_score = f1_av_score/490
tag_36_acc_score = acc_av_score/487

print(tag_36_f1_score)

/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1580: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1580: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1580: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))


0.6742844366861277


In [257]:
tag_36_f1_score

0.6742844366861277

In [261]:
tag_36_acc_score

0.6784381395815248

In [262]:
tag_36_res_prob = res_prob

tag_36_res_prob

1.3538479058987448e-44

# 4 Tag System

In [263]:
tag_set.add('POS')

In [264]:
print(len(tag_set))

36


In [265]:
print(tag_set)

{'TO', 'VBG', 'NN', 'JJS', 'NNS', 'PRP', 'CD', 'JJ', 'EX', 'WP$', 'NNPS', 'VBD', 'WDT', 'FW', 'RBS', 'SYM', 'MD', 'JJR', 'DT', 'POS', 'IN', 'VBZ', 'RBR', 'PDT', 'VBP', 'VB', 'WP', 'LS', 'UH', 'CC', 'RB', 'VBN', 'PRP$', 'WRB', 'NNP', 'RP'}


In [266]:
for i in unfl_data[1]:
  for j in range(len(i)):
    # conversion to Noun
    if(i[j] == 'NN' or i[j] == 'NNS'or i[j] == 'NNP'or i[j] == 'NNPS'or i[j] == 'PRP'or i[j] == 'PRP$'or i[j] == 'WP'or i[j] == 'WP$'):
      i[j] = "N"
    # conversion to Adjective/Adverb
    elif(i[j]=='JJ'or i[j] == 'JJR'or i[j] == 'JJS'or i[j] == 'RB'or i[j] == 'RBR'or i[j] == 'RBS'or i[j] == 'WRB'):
      i[j] = "A"
    # conversion to Verb
    elif(i[j]=='VB'or i[j] == 'VBD'or i[j] == 'VBG'or i[j] == 'VBN'or i[j] == 'VBP'or i[j] == 'VBZ'):
      i[j] = "V"
    else:
      i[j] = "O"

In [267]:
unfl_data[1]

0           [N, N, O, N, A, O, V, O, N, O, O, A, N, N, O]
1                       [N, N, V, N, O, N, N, O, N, V, N]
2       [N, N, O, N, A, O, A, N, O, N, N, N, N, V, V, ...
3       [O, N, O, N, A, V, O, V, N, N, N, V, V, O, A, ...
4       [O, N, N, N, V, A, A, O, N, V, O, N, O, A, A, ...
                              ...                        
3909    [N, A, V, O, A, O, O, N, N, V, V, O, A, N, N, ...
3910    [N, N, V, O, O, A, N, O, N, V, N, O, V, O, N, ...
3911    [N, N, N, V, N, V, O, A, N, O, V, O, N, N, O, ...
3912                                               [N, V]
3913              [N, V, N, V, O, V, N, O, O, A, N, O, A]
Name: 1, Length: 3914, dtype: object

In [268]:
print(unfl_data[0][0])
print(unfl_data[1][0])

Pierre Vinken, 61 years old, will join the board as a nonexecutive director Nov. 29.
['N', 'N', 'O', 'N', 'A', 'O', 'V', 'O', 'N', 'O', 'O', 'A', 'N', 'N', 'O']


In [269]:
# Removing '.' , ',', '`', "'"
# stopwords = ['.',',','`',"'"]

fl_data = []
for i in unfl_data[0]:
  i = i.replace(",","")
  i = i.replace(".","")
  i = i.replace("`","")
  i = i.replace("'","")
  fl_data.append(i)

fl_label = unfl_data[1]

In [270]:
for i in range(10):
  print(fl_data[i])
  print(fl_label[i])

Pierre Vinken 61 years old will join the board as a nonexecutive director Nov 29
['N', 'N', 'O', 'N', 'A', 'O', 'V', 'O', 'N', 'O', 'O', 'A', 'N', 'N', 'O']
Mr Vinken is chairman of Elsevier NV the Dutch publishing group
['N', 'N', 'V', 'N', 'O', 'N', 'N', 'O', 'N', 'V', 'N']
Rudolph Agnew 55 years old and former chairman of Consolidated Gold Fields PLC was named a nonexecutive director of this British industrial conglomerate
['N', 'N', 'O', 'N', 'A', 'O', 'A', 'N', 'O', 'N', 'N', 'N', 'N', 'V', 'V', 'O', 'A', 'N', 'O', 'O', 'A', 'A', 'N']
A form of asbestos once used to make Kent cigarette filters has caused a high percentage of cancer deaths among a group of workers exposed to it more than 30 years ago researchers reported
['O', 'N', 'O', 'N', 'A', 'V', 'O', 'V', 'N', 'N', 'N', 'V', 'V', 'O', 'A', 'N', 'O', 'N', 'N', 'O', 'O', 'N', 'O', 'N', 'V', 'O', 'N', 'A', 'O', 'O', 'N', 'O', 'N', 'V']
The asbestos fiber crocidolite is unusually resilient once it enters the lungs with even brief

In [271]:
from sklearn.model_selection import train_test_split
# X stands for Sentences and Y for the PoS tage associated with them
X = fl_data
Y = fl_label

X_train, X_test, Y_train, Y_test = train_test_split(X,Y,test_size=0.20,shuffle=False)


In [272]:
print(X_train)

['Pierre Vinken 61 years old will join the board as a nonexecutive director Nov 29', 'Mr Vinken is chairman of Elsevier NV the Dutch publishing group', 'Rudolph Agnew 55 years old and former chairman of Consolidated Gold Fields PLC was named a nonexecutive director of this British industrial conglomerate', 'A form of asbestos once used to make Kent cigarette filters has caused a high percentage of cancer deaths among a group of workers exposed to it more than 30 years ago researchers reported', 'The asbestos fiber crocidolite is unusually resilient once it enters the lungs with even brief exposures to it causing symptoms that show up decades later researchers said', 'Lorillard Inc the unit of New York-based Loews Corp that makes Kent cigarettes stopped using crocidolite in its Micronite cigarette filters in 1956', 'Although preliminary findings were reported more than a year ago the latest results appear in todays New England Journal of Medicine a forum likely to bring new attention to

In [273]:
print(len(X_train))
print(len(Y_train))

3131
3131


In [274]:
for i in range(10):
  print(X_train[i]," ",len(X_train[i].split(" ")))
  print(Y_train[i]," ",len(Y_train[i]))

Pierre Vinken 61 years old will join the board as a nonexecutive director Nov 29   15
['N', 'N', 'O', 'N', 'A', 'O', 'V', 'O', 'N', 'O', 'O', 'A', 'N', 'N', 'O']   15
Mr Vinken is chairman of Elsevier NV the Dutch publishing group   11
['N', 'N', 'V', 'N', 'O', 'N', 'N', 'O', 'N', 'V', 'N']   11
Rudolph Agnew 55 years old and former chairman of Consolidated Gold Fields PLC was named a nonexecutive director of this British industrial conglomerate   23
['N', 'N', 'O', 'N', 'A', 'O', 'A', 'N', 'O', 'N', 'N', 'N', 'N', 'V', 'V', 'O', 'A', 'N', 'O', 'O', 'A', 'A', 'N']   23
A form of asbestos once used to make Kent cigarette filters has caused a high percentage of cancer deaths among a group of workers exposed to it more than 30 years ago researchers reported   34
['O', 'N', 'O', 'N', 'A', 'V', 'O', 'V', 'N', 'N', 'N', 'V', 'V', 'O', 'A', 'N', 'O', 'N', 'N', 'O', 'O', 'N', 'O', 'N', 'V', 'O', 'N', 'A', 'O', 'O', 'N', 'O', 'N', 'V']   34
The asbestos fiber crocidolite is unusually resilient 

## Calculation of Probability Matrices

In [275]:
name_set = set()

In [276]:
for i in X_train:
  tokens = i.split(' ')
  for j in tokens:
    name_set.add(j)

In [277]:
len(name_set)

11011

In [278]:
tag_set = set()

for i in Y_train:
  for j in i:
    tag_set.add(j)

In [279]:
print(tag_set)

{'A', 'O', 'V', 'N'}


In [280]:
len(tag_set)

4

In [281]:
df = pd.DataFrame()


df['Names'] = list(name_set)


for i in tag_set:
  df[i] = [1.0 for i in range(1,11012)]

df = df.drop(index = 0)
df.set_index("Names", inplace = True)

df.head(10)

A    O    V    N
Names                           
talk          1.0  1.0  1.0  1.0
853%          1.0  1.0  1.0  1.0
collections   1.0  1.0  1.0  1.0
1901          1.0  1.0  1.0  1.0
psychiatrist  1.0  1.0  1.0  1.0
Al            1.0  1.0  1.0  1.0
laboriously   1.0  1.0  1.0  1.0
twindam       1.0  1.0  1.0  1.0
Cellars       1.0  1.0  1.0  1.0
812%          1.0  1.0  1.0  1.0

In [282]:
df.shape

(11010, 4)

In [283]:
df.loc['coaching','N']

1.0

In [284]:
# Simply traverse all the training set and map and increase the values appropriately in the matrix
for i in range(3131):
  words = X_train[i].split(' ');
  tags = Y_train[i]
  for j in range(len(words)):
    w = words[j];
    t = tags[j];
    # print(w,"->",t)\
    try:
      df.loc[w, t] = df.loc[w,t] + 1
    except:
      print("Exception occured for ",w," ",t);

Exception occured for     O
Exception occured for     O
Exception occured for     O
Exception occured for     O
Exception occured for     O
Exception occured for     O
Exception occured for     O
Exception occured for     O
Exception occured for     O
Exception occured for     O
Exception occured for     O
Exception occured for     O
Exception occured for     O
Exception occured for     O
Exception occured for     O
Exception occured for     O
Exception occured for     O
Exception occured for     O
Exception occured for     O
Exception occured for     O
Exception occured for     O
Exception occured for     O
Exception occured for     O
Exception occured for     O
Exception occured for     O


Found Some Outliers and neglecting them

In [285]:
print(df.head(10))

                A    O    V    N
Names                           
talk          1.0  1.0  6.0  3.0
853%          1.0  2.0  1.0  1.0
collections   1.0  1.0  1.0  3.0
1901          1.0  2.0  1.0  1.0
psychiatrist  1.0  1.0  1.0  2.0
Al            1.0  1.0  1.0  2.0
laboriously   2.0  1.0  1.0  1.0
twindam       1.0  1.0  1.0  2.0
Cellars       1.0  1.0  1.0  2.0
812%          1.0  2.0  1.0  1.0


In [286]:
# Finding sum of each column and then dividing at each column
sum_tags = []
for i in tag_set:
  sum = df[i].sum()
  sum_tags.append(sum)

In [287]:
sum_tags

[18576.0, 33312.0, 21068.0, 36020.0]

In [288]:

for j in range(4):
  for i in range(11010):
    df.iloc[i][j] = float(df.iloc[i][j]/sum_tags[j])


In [289]:
print(df.head(10))

                     A        O         V         N
Names                                              
talk          0.000054  0.00003  0.000285  0.000083
853%          0.000054  0.00006  0.000047  0.000028
collections   0.000054  0.00003  0.000047  0.000083
1901          0.000054  0.00006  0.000047  0.000028
psychiatrist  0.000054  0.00003  0.000047  0.000056
Al            0.000054  0.00003  0.000047  0.000056
laboriously   0.000108  0.00003  0.000047  0.000028
twindam       0.000054  0.00003  0.000047  0.000056
Cellars       0.000054  0.00003  0.000047  0.000056
812%          0.000054  0.00006  0.000047  0.000028


Emission Matrix is ready
Lets create Transmission Matrix Too

In [290]:
df_tp = pd.DataFrame()

tags_row = list(tag_set)
tags_row.insert(0,'S')

tags_cols = list(tag_set)
last = len(tag_set)
tags_cols.insert(last,'E')


# print(tags_row)
# print(tags_cols)
df_tp['Tags'] = tags_row


for i in tags_cols:
  df_tp[i] = [1.0 for i in range(len(tags_cols))]

df_tp.set_index("Tags", inplace = True)

df_tp.shape
df_tp

A    O    V    N    E
Tags                         
S     1.0  1.0  1.0  1.0  1.0
A     1.0  1.0  1.0  1.0  1.0
O     1.0  1.0  1.0  1.0  1.0
V     1.0  1.0  1.0  1.0  1.0
N     1.0  1.0  1.0  1.0  1.0

In [291]:
# Setting up S and E

for i in Y_train:
  start_tag = i[0]
  last_index = len(i)-1
  last_tag = i[last_index]
  try:
    df_tp.loc['S'][start_tag] = df_tp.loc['S'][start_tag] + 1
    df_tp.loc[last_tag]['E'] = df_tp.loc[last_tag]['E'] + 1
  except:
    print("Exception Occured for ", start_tag)
    print("Exception Occured for ", last_tag)

df_tp.head(36) 



A       O     V       N       E
Tags                                     
S     338.0  1499.0  93.0  1205.0     1.0
A       1.0     1.0   1.0     1.0   244.0
O       1.0     1.0   1.0     1.0   332.0
V       1.0     1.0   1.0     1.0   377.0
N       1.0     1.0   1.0     1.0  2182.0

In [292]:
for i in Y_train:
  curr = i
  for j in range(1,len(curr)):
    col = i[j]
    row = i[j-1]
    try:
      df_tp.loc[row][col] = df_tp.loc[row][col] + 1
    except:
      print("Exception Occured!")

In [293]:
tp_sum = df_tp.sum(axis=1)

In [294]:
df_tp

A       O       V       N       E
Tags                                        
S      338.0  1499.0    93.0  1205.0     1.0
A      992.0  1607.0   818.0  3910.0   244.0
O     3317.0  6154.0  2936.0  9593.0   332.0
V     1474.0  4620.0  1473.0  2119.0   377.0
N     1450.0  8452.0  4743.0  8188.0  2182.0

In [295]:
tp_sum = list(tp_sum)

tp_sum

[3136.0, 7571.0, 22332.0, 10063.0, 25015.0]

In [296]:
len(tp_sum)

5

In [297]:
for i in range(5):
  for j in range(5):
    df_tp.iloc[i][j] = df_tp.iloc[i][j]/tp_sum[i]

df_tp

A         O         V         N         E
Tags                                                  
S     0.107781  0.477997  0.029656  0.384247  0.000319
A     0.131026  0.212257  0.108044  0.516444  0.032228
O     0.148531  0.275569  0.131471  0.429563  0.014867
V     0.146477  0.459108  0.146378  0.210573  0.037464
N     0.057965  0.337877  0.189606  0.327324  0.087228

In [298]:
name = df_tp['O'].idxmax()
print(name)

S


>So now both the matrices are ready

>Let's calculate through viterbi algorithm

In [299]:
def viterbi_algo(sentence):
  sentence = sentence.replace(",","")
  sentence = sentence.replace(".","")
  sentence = sentence.replace("`","")
  sentence = sentence.replace("'","")
  sentence = sentence.replace("?","")
  words = sentence.split(" ")

  df_words = pd.DataFrame()
  tag_list = list(tag_set)
  df_words['Tags'] = tag_list
  res_tag_list = []
  for  i in words:
    df_words[i] = [1.0 for i in range(len(tag_set))]

  df_words.set_index("Tags", inplace = True)

  flag = False
  start_word = words[0]
  if(start_word not in name_set):
    flag = True
  for i in range(len(tag_set)):
    if flag == False:
      em_prob = df.loc[start_word][tag_list[i]]
    else:
      em_prob = 1
    t_prob = df_tp.loc['S'][tag_list[i]]
    res_prob = em_prob * t_prob

    # print("Prob for ",start_word,"and Tag: ",tag_list[i],"EmP:",em_prob," TrP:",t_prob," Res: ",res_prob)
    df_words.loc[tag_list[i]][start_word] = df_words.loc[tag_list[i]][start_word] * res_prob
  
  for i in range(1,len(words)):
    flag = False
    # find maximum prob of prev col and tag associated with it. Add the tag into result list
    curr_word = words[i]
    if(curr_word not in name_set):
      flag = True
    prev_col = df_words.columns[i-1]
    last_res_prob = df_words[prev_col].max()
    last_res_tag = df_words[prev_col].idxmax()
    res_tag_list.append(last_res_tag)

    for j in range(len(tag_list)):
      if(flag == False):
        curr_em_prob = df.loc[curr_word][tag_list[j]] 
      else:
        curr_em_prob = 1    
      curr_t_prob = df_tp.loc[last_res_tag][tag_list[j]]
      res_prob = curr_em_prob * curr_t_prob * last_res_prob
      df_words.loc[tag_list[j]][curr_word] = df_words.loc[tag_list[j]][curr_word] * res_prob  

  # print(df_words)
  return [res_prob,words,res_tag_list]

In [300]:
res_prob,words,res_tag_list = viterbi_algo("Sales rose 10% to $13.8 billion from $12.5 billion.")

In [301]:
words

['Sales', 'rose', '10%', 'to', '$138', 'billion', 'from', '$125', 'billion']

In [302]:
res_tag_list

['N', 'V', 'O', 'O', 'O', 'O', 'O', 'O']

> Code is looking to be working fine!

> Let us test this model now

In [303]:
X_test[16]

'The Ginnie Mae 9% securities were yielding 932% to a 12-year average life'

In [304]:
Y_test

3131    [O, A, A, N, V, V, O, O, O, O, V, O, V, O, O, ...
3132    [O, N, N, O, A, N, N, V, A, A, O, O, O, O, N, ...
3133                                                  [N]
3134                    [N, O, O, N, N, V, N, O, O, A, N]
3135                 [A, A, A, N, V, A, O, A, A, O, O, N]
                              ...                        
3909    [N, A, V, O, A, O, O, N, N, V, V, O, A, N, N, ...
3910    [N, N, V, O, O, A, N, O, N, V, N, O, V, O, N, ...
3911    [N, N, N, V, N, V, O, A, N, O, V, O, N, N, O, ...
3912                                               [N, V]
3913              [N, V, N, V, O, V, N, O, O, A, N, O, A]
Name: 1, Length: 783, dtype: object

In [305]:
print(len(X_test))
print(len(Y_test))

783
783


In [306]:
predicted_res_list = []
indices_list = []
for i in range(len(X_test)):
  try:
    res_prob,words,res_tag_list = viterbi_algo(X_test[i])
    indices_list.append(i)
    predicted_res_list.append(res_tag_list)
  except:
    continue
# predicted_res_list

In [307]:
print(len(predicted_res_list))
print(len(indices_list))

490
490


In [308]:
print(indices_list)

[2, 3, 4, 6, 7, 9, 11, 12, 13, 14, 16, 17, 18, 19, 20, 22, 23, 25, 27, 28, 30, 32, 33, 34, 35, 36, 38, 40, 41, 45, 49, 50, 51, 52, 54, 55, 56, 59, 60, 61, 63, 68, 69, 70, 71, 72, 74, 75, 76, 79, 80, 83, 85, 87, 89, 90, 92, 93, 99, 101, 103, 104, 105, 106, 107, 110, 112, 113, 114, 115, 116, 117, 118, 119, 121, 124, 125, 126, 127, 128, 130, 132, 133, 135, 137, 139, 140, 142, 143, 144, 146, 147, 148, 149, 150, 151, 153, 154, 155, 156, 157, 161, 162, 164, 165, 166, 168, 169, 173, 175, 176, 177, 179, 180, 185, 188, 189, 190, 192, 195, 198, 199, 201, 202, 203, 204, 205, 206, 207, 209, 212, 215, 216, 217, 218, 219, 220, 222, 223, 224, 228, 229, 231, 232, 234, 235, 236, 238, 239, 240, 241, 242, 243, 244, 248, 249, 251, 253, 254, 255, 256, 259, 260, 261, 262, 263, 264, 267, 268, 272, 273, 274, 275, 280, 281, 282, 283, 285, 294, 295, 296, 297, 298, 300, 301, 302, 303, 304, 306, 308, 310, 311, 312, 313, 314, 315, 316, 317, 319, 320, 321, 322, 324, 326, 329, 331, 335, 337, 339, 340, 341, 344, 347,

In [309]:
Y_test = Y_test.tolist()

In [310]:
actual_res_list = []
for i in indices_list:
  actual_res_list.append(Y_test[i])

len(actual_res_list)

490

In [311]:
print(predicted_res_list)
print(actual_res_list)

[['N'], ['N', 'O', 'O', 'N', 'N', 'V', 'N', 'O', 'O', 'A', 'N'], ['A', 'A', 'A', 'N', 'V', 'A', 'O', 'O', 'A', 'A', 'A', 'A'], ['N', 'V', 'O', 'N', 'O', 'O', 'O', 'N', 'N', 'N', 'O', 'A', 'N', 'N', 'O'], ['O', 'N', 'N', 'N', 'N', 'O', 'N'], ['O', 'N', 'V', 'O', 'N'], ['N', 'N', 'O', 'N', 'O', 'N', 'V', 'V', 'V', 'O', 'N'], ['N', 'V', 'O', 'A', 'A', 'N', 'N', 'O', 'O', 'N', 'A', 'N', 'V', 'O', 'O', 'O'], ['O'], ['N', 'N', 'V', 'A', 'A', 'O', 'V', 'N', 'O', 'O', 'N'], ['O', 'N', 'N', 'O', 'N', 'V', 'V', 'O', 'O', 'O', 'N', 'N'], ['O', 'V', 'A', 'O', 'N', 'N', 'O', 'O', 'A', 'N'], ['N'], ['N', 'N', 'V', 'A', 'A', 'O', 'O', 'N', 'N', 'O', 'A', 'N', 'N', 'O', 'A', 'A', 'A'], ['O', 'N', 'O', 'N', 'O', 'N', 'N', 'N', 'N', 'O', 'O', 'N', 'N', 'O', 'N', 'O', 'V', 'A', 'V', 'O', 'V', 'A', 'V', 'O', 'O', 'N', 'N', 'N'], ['O', 'N', 'O', 'V', 'O', 'A', 'O', 'O', 'V', 'N', 'O', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A'], ['O', 'N', 'O', 'N', 'O', 'N', 'N', 'N', 'N', 'N', 'N', 'N', 'N', 'N', 'V', 'O', '

In [312]:
# for i in range(490):
#   actual_res_list[i].pop(-1)

In [313]:
for i in range(490):
  if((len(predicted_res_list[i])) != (len(actual_res_list[i]))):
    print("Not Equal")
  else:
    print("Equal with length: ",len(predicted_res_list[i]))
  # print(len(predicted_res_list[i]))
  # print(len(actual_res_list[i]))

Equal with length:  1
Equal with length:  11
Equal with length:  12
Equal with length:  15
Equal with length:  7
Equal with length:  5
Equal with length:  11
Equal with length:  16
Equal with length:  1
Equal with length:  11
Equal with length:  12
Equal with length:  10
Equal with length:  1
Equal with length:  17
Equal with length:  28
Equal with length:  19
Equal with length:  40
Equal with length:  22
Equal with length:  1
Equal with length:  17
Equal with length:  12
Equal with length:  0
Equal with length:  20
Equal with length:  17
Equal with length:  12
Equal with length:  16
Equal with length:  17
Equal with length:  8
Equal with length:  19
Equal with length:  19
Equal with length:  8
Equal with length:  18
Equal with length:  16
Equal with length:  29
Equal with length:  16
Equal with length:  20
Equal with length:  17
Equal with length:  21
Equal with length:  18
Equal with length:  13
Equal with length:  13
Equal with length:  21
Equal with length:  29
Equal with length:  

In [314]:
actual_res_series = pd.Series(actual_res_list)
predicted_res_series = pd.Series(predicted_res_list)

In [315]:
from sklearn.metrics import f1_score
from sklearn.metrics import accuracy_score

from sklearn.preprocessing import MultiLabelBinarizer

actual_res = MultiLabelBinarizer().fit_transform(actual_res_series)
predicted_res = MultiLabelBinarizer().fit_transform(predicted_res_series)

In [316]:
print(actual_res_series[1])
print(predicted_res_series[1])

['N', 'O', 'O', 'N', 'N', 'V', 'N', 'O', 'O', 'A', 'N']
['N', 'O', 'O', 'N', 'N', 'V', 'N', 'O', 'O', 'A', 'N']


In [317]:
count = 0
for i in range(490):
  if(len(actual_res_series[i])!=0):
    count= count+1

count

487

In [318]:
f1_av_score=0
acc_av_score = 0
for i in range(490):
  f1_av_score += f1_score(actual_res_series[i],predicted_res_series[i],average='micro')
  if(len(actual_res_series[i])!=0):
    acc_av_score += accuracy_score(actual_res_series[i],predicted_res_series[i])

tag_4_f1_score = f1_av_score/490
tag_4_acc_score = acc_av_score/487
print(tag_4_f1_score)

/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1580: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1580: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
/usr/local/lib/python3.8/dist-packages/sklearn/metrics/_classification.py:1580: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))


0.7966769465008952


In [319]:
tag_4_f1_score

0.7966769465008952

In [320]:
tag_4_acc_score

0.801584607362297

In [321]:
tag_4_res_prob = res_prob
tag_4_res_prob

5.577964333162506e-39